## Exploring Ebay Car Sales Data

The purpose of this project is to practice using numpy and pandas by cleaning a dataset originally from Kaggle, further dirtied by Dataquest for learning purposes. The dataset is a database of used car listings scraped from Ebay Kleinanzaigan, a classified section of the German Ebay Website. The original data can be obtained from:

https://www.kaggle.com/orgesleka/used-cars-database/data

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv",encoding='Latin-1')

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

A quick glance at the data identifies 20 columns of which they are either the integer type or object type. Some columns contain null values. 

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

Changed columns names from camelcase to snakecase to improve readability.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Some quick observations: nr_of_pictures can be dropped because there are no values. Price and odometer can be converted to numerical values.

In [7]:
autos['price'] = (autos['price']
                   .str.replace('$','')
                   .str.replace(',','')
                   .astype(int)
                 )

In [8]:
autos['odometer'] = (autos['odometer']
                     .str.replace('km','')
                     .str.replace(',','')
                     .astype(int)
                    )

In [9]:
autos = autos.rename(columns={"odometer":'odometer_km'})

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [11]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       abtest vehicle_type  registration_year  gearbox      power_ps  model  \
count   50000        44905       50000.000000    47320  50000.000000  47242   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25756        12859                NaN    36993           NaN   4024   
mean      NaN          NaN        2005.073280      NaN    116.355920    NaN   
std       NaN          NaN         105.712813      NaN    209.216627    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%       NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [12]:
autos['price'].unique().shape

(2357,)

In [13]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [14]:
autos.loc[autos['price'] < 100,'price']

25       90
27        0
30       80
55        1
64       40
71        0
80        0
87        0
99        0
118       0
146       0
167       0
180       0
226       0
234       0
245      50
248       0
259       0
301       0
323       0
327       0
344       1
347       0
366       0
389       0
418       0
427       0
430       0
449       0
487       0
         ..
49082     0
49240     1
49254     0
49301     0
49342     0
49350     0
49377     1
49404     0
49408     0
49479     0
49496     0
49504     0
49507     0
49525     0
49528     1
49538     0
49551     0
49614     0
49728     1
49739     0
49755     0
49793     0
49843    60
49845     1
49880     0
49884     0
49943     0
49960     0
49974     0
49984     0
Name: price, Length: 1762, dtype: int64

After inspecting the values in the price column of the dataset, I noticed that there are outliers within the data. Since this ebay's classified section differs from its other selling format (auction), I removed vehicles listed less than 100 and greater than 400,000 to include reasonable classified ads only. This should remove only about 1,800 records out of 50,000, or 3% of data. 

In [15]:
autos = autos[autos['price'].between(300,400000)]

In [16]:
(autos['date_crawled'].str[:10]
 .value_counts(normalize=True,dropna=False)
 .sort_index()
)

2016-03-05    0.025369
2016-03-06    0.014082
2016-03-07    0.036208
2016-03-08    0.033093
2016-03-09    0.032837
2016-03-10    0.032474
2016-03-11    0.032688
2016-03-12    0.037040
2016-03-13    0.015640
2016-03-14    0.036464
2016-03-15    0.034181
2016-03-16    0.029466
2016-03-17    0.031450
2016-03-18    0.012866
2016-03-19    0.034693
2016-03-20    0.038022
2016-03-21    0.037488
2016-03-22    0.032922
2016-03-23    0.032282
2016-03-24    0.029274
2016-03-25    0.031279
2016-03-26    0.032325
2016-03-27    0.031109
2016-03-28    0.034992
2016-03-29    0.033776
2016-03-30    0.033776
2016-03-31    0.031919
2016-04-01    0.033861
2016-04-02    0.035632
2016-04-03    0.038619
2016-04-04    0.036528
2016-04-05    0.013122
2016-04-06    0.003158
2016-04-07    0.001366
Name: date_crawled, dtype: float64

In [17]:
(autos['ad_created'].str[:10]
 .value_counts(normalize=True,dropna=False)
 .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000043
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000064
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000043
2016-02-05    0.000043
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000043
2016-02-14    0.000043
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000043
2016-02-19    0.000064
2016-02-20    0.000043
2016-02-21    0.000064
                ...   
2016-03-09    0.032986
2016-03-10    0.032154
2016-03-11    0.033008
2016-03-12    0.036870
2016-03-13    0.017027
2016-03-14    0.035077
2016-03-15    0.033925
2016-03-16    0.029978
2016-03-17    0.031087
2016-03-18    0.013506
2016-03-19    0.033584
2016-03-20    0.038128
2016-03-21 

In [18]:
(autos['last_seen'].str[:10]
 .value_counts(normalize=True,dropna=False)
 .sort_index()
)

2016-03-05    0.001067
2016-03-06    0.004267
2016-03-07    0.005355
2016-03-08    0.007126
2016-03-09    0.009580
2016-03-10    0.010455
2016-03-11    0.012333
2016-03-12    0.023940
2016-03-13    0.008855
2016-03-14    0.012439
2016-03-15    0.015725
2016-03-16    0.016152
2016-03-17    0.027887
2016-03-18    0.007254
2016-03-19    0.015512
2016-03-20    0.020654
2016-03-21    0.020462
2016-03-22    0.021486
2016-03-23    0.018563
2016-03-24    0.019544
2016-03-25    0.018904
2016-03-26    0.016600
2016-03-27    0.015554
2016-03-28    0.020910
2016-03-29    0.021870
2016-03-30    0.024452
2016-03-31    0.023833
2016-04-01    0.022894
2016-04-02    0.024814
2016-04-03    0.025092
2016-04-04    0.024473
2016-04-05    0.125757
2016-04-06    0.223287
2016-04-07    0.132905
Name: last_seen, dtype: float64

In [19]:
autos['registration_year'].describe()

count    46868.000000
mean      2004.832850
std         88.065754
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Based on the description of the registration year column, there are values out of realistic year ranges (min and max years). To address this problem, I'm going to assume a realistic registration year range from 1885 - 2016. 1885 was chosen as the minimum as that was the year the first automobile was first created. 

In [20]:
autos.loc[(autos['registration_year'] < 1885) | (autos['registration_year'] > 2016),'registration_year']

10       2017
68       2017
84       2018
113      2017
164      2018
197      2017
253      2017
348      2017
390      2018
438      2017
453      4500
454      2017
457      2017
477      2017
545      2017
548      2017
557      2017
568      2017
577      2017
582      2017
654      2017
740      2017
754      2017
758      2017
765      2017
802      2018
871      2017
889      2017
910      2017
918      2018
         ... 
49130    2017
49153    5000
49154    2017
49178    2018
49185    2019
49218    2018
49245    2017
49259    2017
49261    2017
49262    2018
49266    2017
49283    1001
49343    2017
49347    2017
49354    2018
49389    2017
49411    2018
49522    2017
49557    2017
49561    2017
49653    2017
49662    2017
49689    2017
49696    2017
49731    2017
49770    2018
49796    2017
49841    2017
49910    9000
49935    2017
Name: registration_year, Length: 1840, dtype: int64

Records that are out of this range only account for 1840 records in total. Removing these records will not be detrimental to the overall dataset. 

In [21]:
autos = autos[autos['registration_year'].between(1885,2016)]

In [22]:
autos['registration_year'].value_counts(normalize=True,ascending=True)

1952    0.000022
1943    0.000022
1953    0.000022
1939    0.000022
1938    0.000022
1948    0.000022
1931    0.000022
1950    0.000022
1929    0.000022
1927    0.000022
1941    0.000044
1934    0.000044
1910    0.000044
1951    0.000044
1957    0.000044
1954    0.000044
1955    0.000044
1958    0.000089
1937    0.000089
1956    0.000089
1962    0.000089
1959    0.000133
1961    0.000133
1963    0.000178
1964    0.000267
1965    0.000378
1975    0.000400
1969    0.000422
1960    0.000422
1976    0.000466
          ...   
1985    0.001954
1988    0.002932
1989    0.003620
1990    0.006751
1991    0.006951
1992    0.007395
2015    0.008262
1993    0.008506
1994    0.012548
2014    0.014547
2013    0.017700
1995    0.023119
2016    0.024363
1996    0.027272
2012    0.029026
2010    0.035245
2011    0.035911
1997    0.038643
2009    0.046171
1998    0.048836
2008    0.048992
2007    0.050457
2002    0.054588
2001    0.057631
2006    0.059230
2003    0.059607
2004    0.059807
1999    0.0616

The next step would be to explore the most common car brands listed in these classified ads. Below is my exploration into the average price of the top 20 most common car brands listed. 

In [23]:
top_20 = autos['brand'].value_counts(normalize=True,ascending=False).index[0:20]

In [24]:
average_price={}
for x in top_20:
    x_brand = autos.loc[autos['brand']==x,'price']
    average_price[x] = x_brand.mean()

In [25]:
average_price

{'audi': 9447.399949912347,
 'bmw': 8455.88539814266,
 'citroen': 3886.6692913385828,
 'fiat': 3008.0780269058296,
 'ford': 4022.901682612999,
 'hyundai': 5479.017467248908,
 'mazda': 4256.801169590643,
 'mercedes_benz': 8710.656502242153,
 'mini': 10691.06157635468,
 'nissan': 4957.331864904552,
 'opel': 3172.1037715746857,
 'peugeot': 3222.4580209895053,
 'renault': 2633.496357455075,
 'seat': 4573.17094017094,
 'skoda': 6467.445623342175,
 'smart': 3596.40273556231,
 'sonstige_autos': 13105.930394431554,
 'toyota': 5192.169491525424,
 'volkswagen': 5595.6266428346125,
 'volvo': 5039.0262529832935}

In [26]:
top_20

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

Calculating the average of price of the top 20 most common car brands indicates esentially there are  three tiers of price points for each car brand. 8,000 and up, 5,000 to 8,000, and 5,000 and below. 

In [27]:
average_prices_series = pd.Series(average_price)

In [28]:
average_mileage={}
for x in top_20:
    x_mileage = autos.loc[autos['brand']==x,'odometer_km']
    average_mileage[x] = x_mileage.mean()

In [29]:
average_mileage_series = pd.Series(average_mileage)

In [30]:
average_mileage_series

audi              129189.832206
bmw               132763.287888
citroen           119401.574803
fiat              116412.556054
ford              123927.746618
hyundai           106320.960699
mazda             124181.286550
mercedes_benz     131114.349776
mini               88682.266010
nissan            117621.145374
opel              128923.929256
peugeot           126604.197901
renault           127421.078193
seat              121428.571429
skoda             110988.063660
smart              99734.042553
sonstige_autos     90220.417633
toyota            115991.525424
volkswagen        128570.602460
volvo             138472.553699
dtype: float64

In [31]:
average_prices_series

audi               9447.399950
bmw                8455.885398
citroen            3886.669291
fiat               3008.078027
ford               4022.901683
hyundai            5479.017467
mazda              4256.801170
mercedes_benz      8710.656502
mini              10691.061576
nissan             4957.331865
opel               3172.103772
peugeot            3222.458021
renault            2633.496357
seat               4573.170940
skoda              6467.445623
smart              3596.402736
sonstige_autos    13105.930394
toyota             5192.169492
volkswagen         5595.626643
volvo              5039.026253
dtype: float64

In [33]:
mean_info = pd.DataFrame({'mean_mileage':average_mileage_series,'mean_price':average_prices_series})

In [34]:
mean_info

mean_mileage    mean_price
audi            129189.832206   9447.399950
bmw             132763.287888   8455.885398
citroen         119401.574803   3886.669291
fiat            116412.556054   3008.078027
ford            123927.746618   4022.901683
hyundai         106320.960699   5479.017467
mazda           124181.286550   4256.801170
mercedes_benz   131114.349776   8710.656502
mini             88682.266010  10691.061576
nissan          117621.145374   4957.331865
opel            128923.929256   3172.103772
peugeot         126604.197901   3222.458021
renault         127421.078193   2633.496357
seat            121428.571429   4573.170940
skoda           110988.063660   6467.445623
smart            99734.042553   3596.402736
sonstige_autos   90220.417633  13105.930394
toyota          115991.525424   5192.169492
volkswagen      128570.602460   5595.626643
volvo           138472.553699   5039.026253